torch/nested/_internal/sdpa.py

```diff
@@ -61,6 +61,11 @@ def _validate_sdpa_input(
             f"Expected query, key, and value to all be ragged on the same dimension, but got ragged "
             f"dims {query._ragged_idx}, {key._ragged_idx}, and {value._ragged_idx}, respectively."
         )
+    if query._ragged_idx != 1 and query._ragged_idx != 2:
```

Comment:
What's the purpose of this check? I think it may be too restrictive and breaks if we have multiple batch dims:

e.g. input with shape `(B_0, B_1, S*, num_heads, head_dim)`
transposed to `(B_0, B_1, num_heads, S*, head_dim)` has `_ragged_idx = 3` and semantically we should support this.
```

Reply:
Based on my reading of this function, the NT sdpa code seems unaware of the value of `_ragged_idx`? My intuition was that it would seem difficult to properly handle this correctly without looking at `_ragged_idx`, so I imagined that this must assume the ragged dimension is 2 (or 1 in the case of the special handling at the beginning of the fn).

Is this accurate?

I've removed the check for now though so we can discuss this separately :)

Reply:
> I imagined that this must assume the ragged dimension is 2 (or 1 in the case of the special handling at the beginning of the fn)

The assumption SDPA makes is that the sequence length dim is at `dim=-2` (i.e. the second to last dim). This dim can either be constant or ragged, and it's handled differently between the two cases.

We don't support the embedding dim (`dim=-1`) being ragged, but any other dim can be ragged. If the sequence length dim is not the ragged one, we simply get a "ragged batch dim" or a ragged number of batch elements, which the underlying SDPA kernel doesn't care about.

Reply:
Thanks for the explanation!